translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,355 @@
+// Copyright 2020 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"sort"
+
+	ocmetrics "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes/wrappers"
+	otlpcommon "github.com/open-telemetry/opentelemetry-proto/gen/go/common/v1"
+	otlpmetrics "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	translatorcommon "github.com/open-telemetry/opentelemetry-collector/translator/common"
+)
+
+const (
+	invalidMetricDescriptorType = ocmetrics.MetricDescriptor_Type(-1)
+)
+
+// ResourceMetricsToMetricsData converts metrics from OTLP to internal (OpenCensus) format
+func ResourceMetricsToMetricsData(resourceMetrics *otlpmetrics.ResourceMetrics) consumerdata.MetricsData {
+
+	node, resource := translatorcommon.ResourceToOC(resourceMetrics.Resource)
+	md := consumerdata.MetricsData{
+		Node:     node,
+		Resource: resource,
+	}
+
+	if len(resourceMetrics.Metrics) == 0 {
+		return md
+	}
+
+	metrics := make([]*ocmetrics.Metric, 0, len(resourceMetrics.Metrics))
+	for _, metric := range resourceMetrics.Metrics {
+		metrics = append(metrics, metricToOC(metric))
+	}
+
+	md.Metrics = metrics
+	return md
+}
+
+func metricToOC(metric *otlpmetrics.Metric) *ocmetrics.Metric {
+	labelKeys := allLabelKeys(metric)
+	return &ocmetrics.Metric{
+		MetricDescriptor: descriptorToOC(metric.MetricDescriptor, labelKeys),
+		Timeseries:       dataPointsToTimeseries(metric, labelKeys),
+		Resource:         nil,
+	}
+}
+
+func descriptorToOC(descriptor *otlpmetrics.MetricDescriptor, labelKeys []*ocmetrics.LabelKey) *ocmetrics.MetricDescriptor {
+	if descriptor == nil {
+		return nil
+	}
+
+	return &ocmetrics.MetricDescriptor{
+		Name:        descriptor.Name,
+		Description: descriptor.Description,
+		Unit:        descriptor.Unit,
+		Type:        descriptorTypeToOC(descriptor.Type),
+		LabelKeys:   labelKeys,
+	}
+}
+
+func allLabelKeys(metric *otlpmetrics.Metric) []*ocmetrics.LabelKey {
+	// First, collect a set of unique keys
+	uniqueKeys := make(map[string]struct{}, 0)
```

Comment:
For set is more natural to use `map[string]bool`

Reply:
I prefer to use `struct{}`since it takes 0 bytes, and we never need check for the value (i.e. we never need to add `false` value to the map), also see https://www.davidkaya.com/sets-in-golang/

Reply:
What do you think about naming this:
```suggestion
	allKeys := make(map[string]struct{}, 0)
```